ocsp.go

```diff
@@ -140,117 +168,209 @@ func retryRevocationStatusCheck(totalTimeout *time.Duration, sleepTime time.Dura
 			if err != nil {
-				ocspResponseCacheLock.Lock()
-				delete(ocspResponseCache, encodedCertIDBase64)
-				ocspResponseCacheLock.Unlock()
+				deleteOCSPCache(encodedCertID)
 				return &ocspStatus{
 					code: ocspFailedDecodeResponse,
-					err:  fmt.Errorf("failed to decode OCSP Response value in a cache. CertID: %v", encodedCertIDBase64),
+					err:  fmt.Errorf("failed to decode OCSP Response value in a cache. CertID: %v", encodedCertID),
 				}
 			}
 			ocspRes, err := ocsp.ParseResponse(b, issuer)
 			if err != nil {
-				ocspResponseCacheLock.Lock()
-				delete(ocspResponseCache, encodedCertIDBase64)
-				ocspResponseCacheLock.Unlock()
+				deleteOCSPCache(encodedCertID)
 				return &ocspStatus{
 					code: ocspFailedParseResponse,
-					err:  fmt.Errorf("failed to parse OCSP Respose. CertID: %v", encodedCertIDBase64),
+					err:  fmt.Errorf("failed to parse OCSP Respose. CertID: %v", encodedCertID),
 				}
 			}
 			glog.V(2).Info("using cached OCSP Response")
-			return validateOCSP(encodedCertIDBase64, ocspRes, subject)
+			return validateOCSP(encodedCertID, ocspRes, subject)
 		}
 	}
-	ocspResponseCacheLock.Lock()
-	delete(ocspResponseCache, encodedCertIDBase64) // delete invalid cache entry
-	ocspResponseCacheLock.Unlock()
+	deleteOCSPCache(encodedCertID)
 	return &ocspStatus{
 		code: ocspMissedCache,
-		err:  fmt.Errorf("missed cache. CertID: %v", encodedCertIDBase64),
+		err:  fmt.Errorf("missed cache. CertID: %v", encodedCertID),
 	}
 }
 
-func validateOCSP(encodedCertIDBase64 string, ocspRes *ocsp.Response, subject *x509.Certificate) *ocspStatus {
+func deleteOCSPCache(encodedCertID *certIDKey) {
 	ocspResponseCacheLock.Lock()
-	defer ocspResponseCacheLock.Unlock()
+	delete(ocspResponseCache, *encodedCertID)
+	cacheUpdated = true
+	ocspResponseCacheLock.Unlock()
+}
+
+func validateOCSP(encodedCertID *certIDKey, ocspRes *ocsp.Response, subject *x509.Certificate) *ocspStatus {
 	curTime := time.Now()
+
 	if !isInValidityRange(curTime, ocspRes.ThisUpdate, ocspRes.NextUpdate) {
-		delete(ocspResponseCache, encodedCertIDBase64)
+		deleteOCSPCache(encodedCertID)
 		return &ocspStatus{
 			code: ocspInvalidValidity,
 			err:  fmt.Errorf("invalid validity: producedAt: %v, thisUpdate: %v, nextUpdate: %v", ocspRes.ProducedAt, ocspRes.ThisUpdate, ocspRes.NextUpdate),
 		}
 	}
-	if ocspRes.Status != ocsp.Good {
-		delete(ocspResponseCache, encodedCertIDBase64)
+	switch ocspRes.Status {
+	case ocsp.Good:
 		return &ocspStatus{
-			code: ocspRevokedOrUnknown,
-			err:  fmt.Errorf("bad revocation status. %v: %v, cert: %v", ocspRes.Status, ocspRes.RevocationReason, subject.Subject),
+			code: ocspSuccess,
+			err:  nil,
+		}
+	case ocsp.Revoked:
+		deleteOCSPCache(encodedCertID)
+		return &ocspStatus{
+			code: ocspRevoked,
+			err:  fmt.Errorf("revoked status: reason:%v, at:%v cert: %v", ocspRes.RevocationReason, ocspRes.RevokedAt, subject.Subject),
+		}
+	default:
+		deleteOCSPCache(encodedCertID)
```

Comment:
The OCSP Responses with Unknown cert status should also be allowed to stay in the cache till they expire on their own.

Reply:
ok